pulsar-client/src/main/java/org/apache/pulsar/client/impl/ClientCnx.java

```diff
@@ -1296,6 +1296,9 @@ private void checkRequestTimeout() {
                 break;
             }
             request = requestTimeoutQueue.poll();
```

Comment:
I think we should check the `request` returned by `poll()` is exactly the same with what `peek()` returns. Otherwise, the `request.isTimedOut` call doesn't apply on the removed request.

I think it's better to use `remove` here instead of the null check.

```java
            if (!requestTimeoutQueue.remove(request)) {
                // the request has been removed by another thread
                continue;
            }
```

Reply:
+1

Reply:
`remove()` throw aa `NoSuchElementException` when the queue is empty.

Reply:
```java
            request = requestTimeoutQueue.poll();
            if (request == null) {
                // the request has been removed by another thread
                continue;
            }
```
Wouldn't this be better?

Reply:
> remove() will throw a NoSuchElementException when the queue is empty.

No. I mean the `remove(Object)` API. See the following examples:

```java
        var list = new ConcurrentLinkedQueue<String>();
        list.add("hello");
        var s = list.peek();
        list.clear(); // assuming another thread remove the value
        System.out.println(list.isEmpty()); // true
        boolean removed = list.remove(s); // no exception would be thrown
        System.out.println(removed); // false
```

Reply:
My bad, I misread the API.